### PR TITLE
Fixes to avoid getting nodes cache into invalid state

### PR DESCRIPTION
**Pull request by saskliutas**
https://github.com/iTwin/imodel-native/issues/750 issue is difficult to repeatedly reproduce but managed to spot some problems that could lead to backend hang when nodes cache ends up in invalid state. Made some changes to avoid that:

- Always check for locked children levels when locking hierarchy level. Managed to end up in a situation where parent and child hierarchy levels were lock and updated at the same time. This change also allowed to remove additional locks held during hierarchy update after iModel change.
- Moved hierarchy level caching in `SameLabelGroupingNodesPostProcessorDeprecated` to be done in a transaction after lock is aquired. Also added additional check before caching hierarchy level and data source in case they were cached while checking for child nodes.

### CODE REVIEW

File: iModelCore/ECPresentation/Source/UpdateHandler.cpp [hunk 1]
```diff
@@ -596,6 +557,7 @@ struct HierarchyRefresher
 
         AffectedHierarchyLevelIdentifier grandparentLevel(parentHierarchyLevelIdentifier, grandParent.IsValid() ? grandParent->GetKey().get() : nullptr);
 
+        newProvider.GetContext().GetHierarchyLevelLocker().Unlock();
         std::shared_ptr<IHierarchyLevelLocker> grandParentHierarchyLock = m_nodesCache->CreateHierarchyLevelLocker(grandparentLevel.GetHierarchyLevelIdentifier(), true);
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/UpdateHandler.cpp:561]: Do we really need to lock here? It looks like a lock gets created by `RefreshHierarchyLevel` -> `RefreshNavNodesProvider`

File: iModelCore/ECPresentation/Source/Hierarchies/NavNodeProviders.cpp [hunk 2]
```diff
@@ -3538,10 +3558,9 @@ NavNodesProviderPtr SameLabelGroupingNodesPostProcessorDeprecated::_PostProcess(
 
     // attempt to find cached merged nodes provider - success means the whole hierarchy level is already post-processed and in cache.
     // it's more efficient to use the cached version compared to loading and merging everything again, so just return the cached provider.
-    DataSourceIdentifier mergedDatasourceIdentifier(GetHierarchyLevelIdentifier(*context).GetId(), {}, context->GetInstanceFilterPtr());
-    mergedDatasourceIdentifier.SetResultSetSizeLimit(context->GetResultSetSizeLimit());
-    DataSourceInfo mergedDatasourceInfo = context->GetNodesCache().FindDataSource(mergedDatasourceIdentifier, context->GetRulesetVariables());
-    if (mergedDatasourceInfo.GetIdentifier().IsValid())
+    HierarchyLevelIdentifier hlIdentifier = GetHierarchyLevelIdentifier(*context);
+    DataSourceInfo mergedDatasourceInfo = GetMergedDataSourceInfo(*context, hlIdentifier.GetId());
+    if (mergedDatasourceInfo.GetIdentifier().IsValid() && mergedDatasourceInfo.IsInitialized())
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/Hierarchies/NavNodeProviders.cpp:3563]: Is it possible that `mergedDatasourceInfo.IsInitialized()` returns `true`, but `mergedDatasourceInfo.GetIdentifier().IsValid()` - not? Maybe only one check is enough?

File: iModelCore/ECPresentation/Source/UpdateHandler.cpp (hunk 1 above)
Comment (saskliutas) [reply iModelCore/ECPresentation/Source/UpdateHandler.cpp]: We don't need. Removed

File: iModelCore/ECPresentation/Source/Hierarchies/NavNodeProviders.cpp (hunk 2 above)
Comment (saskliutas) [reply iModelCore/ECPresentation/Source/Hierarchies/NavNodeProviders.cpp]: Yes, it's enough to check `IsInitialized()`

**Review by grigasp — APPROVED** [review]

**Comment by grigasp** [comment]
@mergifyio backport release/4.x

**Pull request merged**